transcom/mymove
# B 18270 int - Weight Estimator Excel file to PDF conversion

## [B-18270](https://www13.v1host.com/USTRANSCOM38/story.mvc/Summary?oidToken=Story%3A873141)

## Summary

When Customers don't have a weight receipt to upload they can use the official Weight Estimator excel spreadsheet as their upload for the weight receipt. We need this spreadsheet to be a part of the payment packet which is made up of PDF files. We use the Excelize library to parse the massive Weight Estimator .xlsx file and then convert the parsed data to JSON and use that to populate the newly created Weight Estimator PDF template file so we can then finally add it to the payment packet when it is generated. 

Is there anything you would like reviewers to give additional scrutiny?

Don't forget to review the weightTicketParser files, they are new files and have a lot of changes so github doesn't expand them automatically.

## Verification Steps for the Author

These are to be checked by the author.


## Verification Steps for Reviewers

These are to be checked by a reviewer.


### Setup to Run the Code

- [Instructions for starting storybook](https://transcom.github.io/mymove-docs/docs/frontend/setup/storybook)
- [Instructions for starting the MilMove application](https://transcom.github.io/mymove-docs/docs/getting-started/application-setup/)
- [Instructions for running tests](https://transcom.github.io/mymove-docs/docs/getting-started/development/testing)

### How to test

To test this you will need to upload a Weight Estimator spreadsheet with data in it. You can use this one which is the same one that was put into the testdata folder and the tests will use.
[Weight Estimator Full.xlsx](https://github.com/transcom/mymove/files/14935713/Weight.Estimator.Full.xlsx)
Alternatively you can go grab a blank copy of the [Weight Estimator spreadsheet](https://www.ustranscom.mil/dp3/weightestimator.cfm) tool and fill it out and upload that. 

1. Login as a new Customer and create a new PPM shipment and submit it
2. Login as Service Counselor and approve the shipment
3. Login as a Customer and click the Upload PPM Documents button
4. Fill out the data until you get to the Upload section for Weight Tickets
5. Upload your Weight Estimator spreadsheet for either the Empty Weight or Full Weight tickets.
6. The file should be parsed successfully and you will see your filename with a .pdf extension displayed in the upload section. 
7. Now to generate a Payment Packet to verify the data was parsed correctly.
8. Easiest way is to grab the PPM shipment ID, right click in Chrome and Inspect, go to Network tab, refresh the screen if Network tab is empty, then find mto_shipments click on it, then look for ppm_shipment and expand that and copy the "id"
9. Go to the internal [API swagger page] (http://milmovelocal:3000/swagger-ui/internal.html)
10. Find the payment-packet endpoint ![image](https://github.com/transcom/mymove/assets/149000222/b4322987-7dd8-44ed-af4a-efa325a6e4e2)
11. Click the Try It Out button and then enter your PPM Shipment ID and click Execute
12. You should get a 200 response and a message about not being able to read the data, that is expected.
13. Copy the URL you are given below the curl statement, if it won't let you copy it right away give it a few seconds to finish trying to parse the response data then try to copy the link again.
14. Open that link in the browser and you should see a Payment Packet, look for the Weight Ticket pages and compare the data to the data in your uploaded .xlsx file

### Frontend

508 compliance](https://www.ssa.gov/accessibility/andi/help/install.html).

### Backend


## Screenshots
![image](https://github.com/transcom/mymove/assets/149000222/f30d1857-ad40-4699-875a-8d5fcc6fbf2e)
